### ISSUE TITLE
polymer.js:11079 Uncaught ReferenceError: PolymerExpressions is not defined

### ISSUE THREAD

**Issue by reggi** [Polymer/polymer#1238]
This is one of those errors that only comes about when I compile my own version of polymer with browserify or similar module loaders.

```
polymer.js:11079 Uncaught ReferenceError: PolymerExpressions is not defined
```

**Comment by davidmaxwaterman** [comment]
I'm unclear why you're using browserify. Innm, browserify is for using node modules in a browser. Polymer already targets the browser, so it isn't appropriate to use browserify - you just use a <link rel="import" href="path-to-polymer.html"> element in your index.html or other html file.

**Comment by reggi** [comment]
Hey @davidmaxwaterman, thanks for your comments (including https://github.com/Polymer/polymer/issues/1236#issuecomment-76690902)!

I think that my goal or what I'd like to see is an easier way to _build_ components with _dependencies_. You're absolutely right, you mentioned in your other comment it doesn't _require_ a compile step. I have a [working demo](https://github.com/reggi/demo-browserify-polymer) here of two components using the same dependency `moment`. And your right In these examples, I'm not [importing polymer using the link tag](https://www.polymer-project.org/platform/html-imports.html) or [relying on `Polymer` being a global variable](https://github.com/addyosmani/polymer-browserify-vulcanize/blob/master/app/scripts/my-element.js). All I can really share is my desire to make things like this, and if the community disagrees then so be it.  I just feel like the polymer project is a little lacking when it comes to minifying, packaging, and dependencies, which are all real-world issues. 

```
var Polymer = require("polymer")
var moment = require("moment")

Polymer("moment-time", {
  ready: function() {
    var $this = this
    setInterval(function(){
      $this.time = moment().format($this.format)
    }, 1000);
  }
});
```

**Comment by davidmaxwaterman** [comment]
OK, I see what you're getting at. I'm not sure about that objective, so I'll leave it to others to answer, and be interested in reading what is said too :)

**Comment by davidmaxwaterman** [comment]
Sorry to come back to such an old thread, but I thought I'd mention a method of loading dependencies in a [require.js](http://requirejs.org/) fashion - though actually with [System.js](https://github.com/systemjs/systemjs). I't's not quite as succinct as Require.js, but it does the job. If you can use ES2015 destructuring, then it might be a lot more like Require.js does it.

We implemented it here (a 0.5 app) :

<https://github.com/solettaproject/soletta-visual-editor/blob/master/app/elements/slv-editor.html#L178>

```
  window.slvEditorReady = Promise.all(
  [
    System.import('Component'),
    System.import('FactoryDecorator'),
    System.import('GraphQuery'),
    System.import('ComponentSimilarGrouper'),
    System.import('NodePlacer'),
    System.import('lodash'),
    System.import('js-signals'),
    window.slvComponentSelectPopupReady
  ])
  .then(function (imports) {
    'use strict';
    var Component = imports.shift();
    var FactoryDecorator = imports.shift();
    var GraphQuery = imports.shift();
    var ComponentSimilarGrouper = imports.shift();
    var NodePlacer = imports.shift();
    var _ = imports.shift();
    var signals = imports.shift();
```

Having said this, I had a discussion with @sjmiles who knows infinitely more about this sort of thing than I do, and he said it really isn't needed - I never did understand what he was saying though.

I suppose there are two questions I'd like to know the answer to :

1) The encapsulation of the Polymer() call inside an asynchronous callback function, meaning that, by the time the JavaScript in that imported file has finished running, the Polymer() call may not have been run (it's dependencies are loaded first, often meaning delays due to disk and/or network), so elements that import it, can't rely on the element having been registered/etc. How to solve that problem?

2) How to 'rename' a dependency? This is done by changing the name of the parameter in the require() callback. I'm not sure how you would do this with an html import.

I've seen some mention of things that might be relevant to the above at the Polymer Summit....yes, from @ebidel 's talk, he mentioned a `beforeRegister` method which sounded like it might fit into this somewhere, but exactly I'm not sure where or how.

@robdodson said he'd also come up with some similar solution to the promise based one above, but I hope Scott or Eric can elaborate on the issues, and detail their recommended solutions.
